### PR TITLE
NODE-446 /node/status returns state height that lags behind blockchai…

### PR DESCRIPTION
**Pull request by petermz**
JIRA: https://wavesplatform.atlassian.net/browse/NODE-446

Instead of reading persisted state height from StateWriter, I'm now reading complete (persisted + in-memory) state height from SnapshotStateReader

### CODE REVIEW

File: src/main/scala/com/wavesplatform/state2/StateStorage.scala [hunk 1]
```diff
@@ -17,9 +17,14 @@ class StateStorage private(file: Option[File]) extends VariablesStorage(createMV
 
   override protected val Version = 2
 
+  @volatile private[state2] var lastUpdated: Long = 0L
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state2/StateStorage.scala]: Please clean this up, make a public getter and private setter, package-private vars is a mess

File: src/main/scala/com/wavesplatform/state2/StateWriter.scala [hunk 2]
```diff
@@ -5,22 +5,18 @@ import java.util.concurrent.locks.ReentrantReadWriteLock
...
 import scorex.utils.ScorexLogging
-import scorex.utils.Synchronized.ReadLock
 
 trait StateWriter {
   def applyBlockDiff(blockDiff: BlockDiff): Unit
 
   def clear(): Unit
 
-  def status: Observable[Status]
+  def lastUpdated: Long
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state2/StateWriter.scala]: is `Status` not used anymore?

File: src/main/scala/com/wavesplatform/state2/StateStorage.scala [hunk 3]
```diff
@@ -17,9 +17,14 @@ class StateStorage private(file: Option[File]) extends VariablesStorage(createMV
...
+
   def getHeight: Int = getInt(heightKey).getOrElse(0)
 
-  def setHeight(i: Int): Unit = putInt(heightKey, i)
+  def setHeight(i: Int): Unit = {
+    putInt(heightKey, i)
+    lastUpdated = System.currentTimeMillis
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/state2/StateStorage.scala]: Why not use `Time`? It is not critical here, but it's best if we use single source of time everywhere

File: src/main/scala/com/wavesplatform/http/NodeApiRoute.scala [hunk 4]
```diff
@@ -49,7 +49,7 @@ case class NodeApiRoute(settings: RestAPISettings, blockchainUpdater: Blockchain
   @ApiOperation(value = "Status", notes = "Get status of the running core", httpMethod = "GET")
   def status: Route = (get & path("status")) {
     val (bcHeight, bcTime) = lastHeight().map { case LastBlockInfo(_, h, _, _, t) => (h, t) }.getOrElse((0, 0L))
-    val (stHeight, stTime) = lastState().map { case StateWriter.Status(h, t) => (h, t) }.getOrElse((0, 0L))
+    val (stHeight, stTime) = (stateReader().height, stateWriter.lastUpdated)
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/http/NodeApiRoute.scala]: There's th `trait BlockchainDebugInfo`, please use it this stuff, BlockchainUpdater and StateWriter shouldn't be here at all

**Comment by petermz** [comment]
In this commit:
- Moved stuff into BlockchainDebugInfo to avoid exposing StateReader/Writer
- Added history height and timestamps to StateDebugInfo (should I rename this class now?)
- Time is used everywhere instead of System.currentTimeMillis

**Comment by petermz** [comment]
No, wait, that doesn't work because of locking =(

**Comment by ismagin** [comment]
Looks like all you need to expose from not-yet-rebuilt state is height. Maybe you can just expose coeval instead of  accepting callback?

**Comment by ismagin** [comment]
NODE-467 Stop exposing internal entities to NodeApiRoute #780 suites better

**Pull request closed without merging**